### ISSUE TITLE
Switch ui-chart from ChartJS to Apache E-Charts

### ISSUE THREAD

**Issue by joepavitt** [FlowFuse/node-red-dashboard#782]
### Description

[Apache ECharts](https://echarts.apache.org/en/index.html) offer a far vaster collection of charts, with a very similar API, and [extensive examples in their documentation](https://echarts.apache.org/examples/en/index.html).

Whilst @Steve-Mcl mentioned this a while ago, we had no obvious reason to be investing the time in switching these over, _however_, with https://github.com/FlowFuse/node-red-dashboard/issues/676 being raised and investigated, I feel we now have no other option but to investigate Apache eCharts as an alternative charting library as the performance of ChartsJS, with regularly updating charts is not viable.

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

**Comment by nileio** [comment]
most of my charts are apache echarts - my suggestion is to have a less priority to this feature because echarts is a real beast with every possible option and chart type best implemented in `ui-template` in my opinion .. it is however the best out there

**Comment by joepavitt** [comment]
> my suggestion is to have a less priority to this feature because echarts is a real beast with every possible option

We wouldn't be exposing _every_ option in the `ui-chart` node, we don't know vs. what ChartJS actually offers. We do though, need a more reliable charting solution that can handle the quanities (and frequency) of data our dashboards are providing.

**Comment by nileio** [comment]
hands down, I vouch for Apache eCharts as in my own implementation I actually hardly ever used Node-RED own charts widgets and depend on eCharts for everything- however, i do have to construct echarts so-called `option` which can be quite complex and so what I meant to say is exposing echarts  `option` as attributes for each chart type widget in Node-RED would be rather a complex and long exercise - I think alternatively we could work on a dedicated widget for echarts which simply takes in a JSON object for its `option` and let the user construct it anyway they like - this would be easier to implement and would provide flexibility.. however this depends on whether rather include echarts as an integral part of the project or leave it to community contributions..

**Comment by joepavitt** [comment]
More evidence we should be looking into eCharts: https://discourse.nodered.org/t/chart-performance-issues-even-with-very-few-points/89672/7?u=joepavitt

**Comment by bartbutenaers** [comment]
FYI In my local development environment I have a ui-chart node now running on top of Echarts instead of ChartJs.  But there is still a _**LOT**_ of work to be done before it is completed.

**Comment by bartbutenaers** [comment]
Hmm, would appreciate to get some "constructive" feedback because I am about to implement a ***breaking change***.
When you create e.g. a line chart, you can select how each datapoint needs to look like:

![Image](https://github.com/user-attachments/assets/171a2cb7-5484-47dc-bccd-b411bb6312fc)

However Echarts does only have some of those [symbols](https://apache.github.io/echarts-handbook/en/how-to/chart-types/scatter/basic-scatter/#symbol-style) available:
+ circle
+ rect'(rectangle)
+ roundRect'(rounded rectangle)
+ triangle
+ diamond
+ pin
+ arrow

I am tempted to simply replace the old dropdown with the new values.  But that means that existing ui-chart nodes (that have one of the unsupported symbols selected) will get a new ***default symbol***.  Which imho is not the biggest world problem, but just asking seen the kind of feedback we receive in this repo recently...

In case the missing custom markers are really required for some reason, I only see two solutions:

1. Create custom symbols.  In that case the symbol name wouldn't be one of the above names, but instead a custom SVG string ( for example `path://M51.911,16.24...C`).  However I see a few disadvantages of that approach:
   + Performance.  One of the main reasons to switch from ChartJs to ECharts is speed.  However I have no idea how speed is impacted when you use a (larger) SVG string for each point.  Especially since an Echart chart is no `svg` element, but instead they draw on top of a `canvas` element.  So I assume they somehow need to convert the SVG to draw it on a canvas.  Doesn't smell like fast too me...
   + Pollution of the ui-chart node code with all those SVG strings.
   + Extra work.  If we decide to go the custom svg approach, I hope that at least somebody can contribute some 'compact' svg strings for all the missing symbol types...
2. Allow the user to enter a custom symbol somehow, i.e. let him enter a custom svg string.  But this has similar disadavantages as above.  Of course such an option could also be added later on in the future...

Thanks for reading!
Bart

**Comment by hotNipi** [comment]
With SVG you can easily enough fall into the problems with iOs older versions (which still are good enough for wall display purposes ) Mostly those issues can be fixed but as the standards are not followed correctly it's case by case to investigate. 
So you know ...

**Comment by joepavitt** [comment]
> I am tempted to simply replace the old dropdown with the new values

That's the right course of action I'd say

**Comment by joepavitt** [comment]
Before diving into this heavy (which tiu may already have) its worth triple checking performance improvements, e.g.

How does a Dashboard Chart with ChartJS perform vs. Dashboard Chart with ECharts

**Comment by bartbutenaers** [comment]
@hotNipi @joepavitt 
I appreciate your input guys.

Then I simply go for the EChart symbols.

I try to have a look at performance all the time while refactoring this node.  But doing a side by side comparison is ***way*** out of my free time limits.  I thought that it was already clear from the above posts that Echarts was worth the effort of doing all of this.  Hopefully then it was worth all the efforts at the end...

**Comment by bartbutenaers** [comment]
> doing a side by side comparison is way out of my free time limits

But anyway if Echarts would be slower compared to ChartJs, it would be useless to continue with my refactoring.  So I have ***quickly*** done a very basic performance test, to know whether I should continue or not:

I have created a function node to inject 10000 datapoints into a line chart, and repeated that test a series of time:
+ For the original ***ChartJs*** ui node, it took about ***10 seconds***.  The resulting chart was rather blurry and not very responsive.
+ For the refactored ***Echarts*** ui node, it took about ***2 seconds***.  The resulting chart looks rather good and is still responsive:

![Image](https://github.com/user-attachments/assets/ffa9524d-6fe3-45d1-a502-569e67c94f0c)

In both tests I have commented the function to limit the data points, to make sure that isn't interfering in some way.

Of course this isn't really very accurate, because perhaps there is some performance bottleneck in the original ChartJs node (so perhaps the slow performance isn't caused by ChartJs itself).  Or perhaps I have unaccidentally removed something slow from the code, that needs to be added again afterwards.  Don't know.  Unfortunately I don't have time to do more in depth tests, on top of everything else...

However based on this result, I think it is worth the effort to refactor the chart node, and bite the bullet...

**Comment by joepavitt** [comment]
> So I have quickly done a very basic performance test, to know whether I should continue or not:

That's ideal @bartbutenaers and exactly what I wanted to sanity check against. Didn't want you investing all this time for us to have got something wrong re: assessing performance along the way

**Comment by bartbutenaers** [comment]
@joepavitt 
Yes but very good that you mentioned it. Would be awful if all my refactorings would have been for the garbage bin.  Perhaps I will do another quick test tonight: instead of replacing 10000 datapoints at once, simply append 10000 individual datapoints in a sequence.  Because in ChartJs you could upload the data, while in the recent Echarts versions you need to pass him/her/it a big object containing both the data and the config.  Then Echarts itself determines the delta to see what has changed.  Which might be a lot of overhead when you only have updated a single data point.  

Although not very sure if that would be a party stopper?  Because not sure if that is a realistic use case: I assume that in most cases of such large number of datapoints injected into a short time interval, that those datapoints are e.g. queried from a timeseries db and injected at once in the chart to visualize them.  So that in those cases people use replace instead of append.  Not sure.  Anyway I will test and share the results here.

BTW I think the line chart is completely migrated, since all features 'seem' to be working fine in my local setup.  All other chart types of the old ui chart node still need to be done...

**Comment by bartbutenaers** [comment]
> Echarts itself determines the delta to see what has changed. Which might be a lot of overhead when you only have updated a single data point.

As I expected, periodically adding new datapoints to a chart (which contains a large number of datapoints already) by calling `setOption` over and over again, is ***horrible slow***.  While ChartJs handles this rather well, Echarts becomes slower and slower when the number of datapoints in the chart is increasing.  At the end the browser becomes completely unresponsive.  I assume that happens because - when you call `setOption` for every new datapoint - Echarts will calculate the delta by itself.  Which means comparing a large amount of data that hasn't been changed, only to find at the end that 1 single datapoint has changed...

Fortunately Echarts also has an `appendData` function, for this kind of use cases.  So I have refactored the code to use that one in those cases, which means a bit of code pollution unfortunately.  But did my best to keep to code readable.  Then I was stuck because the appended data points are NOT rendered.    After creating a [ticket](https://github.com/apache/echarts/issues/20734) in their repo, it seemed that you need to trigger the rendering manually.

When I call `chart.resize()` after each `chart.appendData(datapoint)` call it seems to be working fine.  We can also optimize this further afterwards, by e.g. calling the resize less often (e.g. once a second).  Then I get these results:

| Points  |ChartJs | Echarts |
| ------- | ------- | -------- | 
| 500 | 7 sec |  8 sec |
| 1000 | 18 sec |  20 sec |
| 2000 | 65 sec | 37 sec |

As a result for smaller number of datapoints, ChartJs is a bit faster.  But when the number of datapoints increases, Echarts seems to become much faster...

So I think we can conclude that Echarts is faster compared to ChartJs, in both situations:
+ When the entire large dataset is being ***replaced*** at once.
+ When datapoints are being ***appended*** one by one to a large dataset.

Enough performance stuff, I will now continue my migration of the ui-chart node to Echarts...

**Comment by joepavitt** [comment]
37 seconds to load a chart with only 2k data points still feels abnormally slow to me though?

**Comment by hotNipi** [comment]
Side note:
Users should not have the illusion that the system can handle no matter how much they push it. 
I don't have any good idea how to make it clear, where and how to explain it. But without that the overall outcome is pretty much same.

**Comment by bartbutenaers** [comment]
@hotNipi 
Yes completely true.  I have seen feature requests that the the ui-chart node should be able to coope with  huge amounts of datapoints.  While that indeed is a nice to have a single node out of the box that does everything, the main purpose of a chart node is to display datapoints that you deliver to it.  It is not a timeseries database node, it is not a downsampling node, ...  There are plenty of other Node-RED nodes available that do such things wonderful. 

Anyway if anybody else wants to have that, they are free to contribute those features the code.  But I have no interest/time/energy for it...

**Comment by bartbutenaers** [comment]
> that - in case of appending data - the refresh rate should be adjustable in the config screen. For example once every N seconds, where N can be 1.

I have implemented this feature, but the chart updating looks very ugly.  No idea what I am doing wrong.  I have added a [feature request](https://github.com/apache/echarts/issues/20739) in the Echarts repo.  I am pretty sure I am overlooking an existing feature that can do the job, but cannot find anything related unfortunately.

**Comment by bartbutenaers** [comment]
The people from Echarts have identified my issue as a _**bug**_ in their line charts, because it works fine for their scatter charts.
Hopefully they are able to fix it one day...
However they also provided me with a decent _**workaround**_, which I will try to implement tonight.
I have to admit that their support is very constructive.
To be continued.  Now I am off to the daily job...

**Comment by bartbutenaers** [comment]
Ok, I think I finally have most of the _**existing functionality**_ migrated to Echarts.  That was a lot of work.
I have also done a lot of refactoring, in an attempt to make it easier to add new chart types/features in the future.

What I still need **_TODO_**:
+ Refactor the code a second to make sure it is a bit consistent.
+ Solve all the TODO's in the code that I have added during the migration.
+ Add some basic dynamic property via `msg.ui_update` perhaps, or at least make sure that changes in the config screen afterwards are being used in the dashboard.

Pff continiously refactoring existing code, is not really relaxing after a heavy day at the daily job...
Although I had told myself not to add new features, I really wanted to do something fun for 30 minutes.
I had always wanted to add a ***zoom*** feature for my charts in my own dashboard, so I added this to the config screen:

![Image](https://github.com/user-attachments/assets/7266e546-893a-4cb2-95e2-1d5579131ae1)

And yes, imho that looks very cool:

![Image](https://github.com/user-attachments/assets/ba80d356-7057-499b-8855-1d371fe108eb)

I won't be adding other new extra features in this version.
Unless I am getting bored migrating existing code, then perhaps I might add something else (for my own home automation).
Now back to the boring stuff.

**Comment by bartbutenaers** [comment]
> Unless I am getting bored migrating existing code, then perhaps I might add something else (for my own home automation).

Ok I had no energy left tonight to start refactoring again.  And I needed this also for my own dashboard:

![Image](https://github.com/user-attachments/assets/6f1bd1f5-e7cc-4e4f-b041-c14388193f16)

No fancy color gradients, just let Echarts calculate its own fill color (based on the line color).  If anybody else wants to have more exotic features, then are free to add those in the future.  But for me this is a good enough for now:

![Image](https://github.com/user-attachments/assets/ce003197-6752-4413-ab23-dcc8ba23a1c3)

My time is up for today, will continue tomorrow with the boring stuff...

**Comment by bartbutenaers** [comment]
I had implemented the workaround from the Echarts team, but now it looks even slower as before.  Don't understand it yet.
So I did some ***profiling*** tonight to find the root cause, and the results are completely not what I expected:

![Image](https://github.com/user-attachments/assets/c07b4886-c1fb-4623-bd0f-3cf6ef697546)

1. So the Echarts `resize` (to trigger the re-rendering to show the newly appended datapoints, is only using very minimal cpu time.  However that might be the case because I now have setup my new throttling parameter to only refresh every 0.5 seconds.  So that might become higher when I reduce my refresh throttling.
2. I also loose ***18.7%*** cpu time on the `getOption` to get the Echarts config every time a new datapoint arrives.  That is indeed something that could be reduced a lot because it makes no sense to load that over and over again, if nothing has changed.  So that really needs to be refactored.
3. But ***68.2%*** of the cpu time is lost in this code snippet:
   ```
   this.$store.commit('data/restrict', {
                widgetId: this.id,
                min: cutoff,
                points
            })
   ```

I have no clue at the moment what that code is doing, it is a leftover from the original ui-chart node with ChartJs.  As mentioned before I find the datastore setup quite hard to understand.  So Echarts isn't the  performance bottleneck at all, but it seems to be this stuff that is a party stopper.

Would be nice if somebody could enlighten me how this works, and if I am allowed to remove that code.

**Comment by joepavitt** [comment]
From memory, and without diving back into the code, that's the piece that removes older data points, I think.

I think `cutoff` will likely be a timestamp, and the `restrict` action, runs against the data store, with older data points being cleared.

**Comment by bartbutenaers** [comment]
@joepavitt,
Thanks for the quick response.  Ok that makes sense.  It [seems](https://github.com/FlowFuse/node-red-dashboard/blob/main/ui/src/store/data.mjs#L56) indeed to remove the older points in the data store.  Then I should also try to call it only per batch of N datapoints (similar to the re-rendering) and perhaps have a look whether there is a faster way compared to slicing arrays.

**Comment by bartbutenaers** [comment]
> whether I might have damaged this way the working of the Vuex datastore? 

I see in theVueJs [documentation](https://vuejs.org/guide/essentials/list.html#mutation-methods) that they wrap the standard ´splice´ function to make sure their reactivity magic is called. 

So I can keep my fix in the datastore code, which is a huge performance improvement (when dealing which large arrays).

**Comment by bartbutenaers** [comment]
In my previous test I could inject 2000 separate datapoints via separate input messages, and process those messages and render them in the line chart in ***37*** seconds.

Now with the `splice` datastore fix and when I set the refresh/rerender interval to 300 msec (which is still rather smooth for my eye) I can do this in ***21*** seconds.  However of course when the dataset grows further, the performance will decrease...

Currently the profiling shows - after a quick view - that Echarts `resize` is consuming most CPU, so now it is mostly charting stuff.   Which is kind of normal.  Of course there might be Echarts settings that allow to improve performance further.

In case anybody has higher demands and all of this isn't fast enough for their use case, I recommend these options:

1. Start profiling to locate the bottlenecks and create a pull request to improve the code further.   Others will be grateful for that...
2. Redesign your Node-RED flows.  Ingesting - without any filtering - datapoints continiously at higher rates that you can visualize decently, seems very weird to me. Of course it is nice if you can simply use a chart as a timeseries database, so that you can create a one-node flow in 30 seconds.  But a ui-chart node isn't a high-performant timeseries database, and in such use case you will need to start looking at your flows from a performance point of view.  

This is as far as I go for performance.   I need to start working soon on my own video surveillance stuff at home. 
I will try to finalize my other todo's for this migration, but that is about it...

**Comment by gstout52** [comment]
Moving back to 2.23 because of limited engineering resources.

**Comment by joepavitt** [comment]
Just a quick update to say I've started this today. Initial sweep and API updates are going well. Complex areas are going to be around data updates and re-writing the histogram charting we had, as that had a lot of custom work put into it. 

I am likely to split the Histogram into it's own component in order to split that logic as it bloats the file a lot.

Whilst here, I will also be keeping an eye on general tidying I can do, as this widget has grown a lot with many edge cases accounted for. Priority is getting existing charts migrated though. Clean up will follow.

**Comment by joepavitt** [comment]
~Further Update: All charts are now migrated and working, except for one edge case - multi-dimensional pie charts. These were supported out of the box (for sizing the charts and nesting them) in ChartJS, this is not the case in eCharts and requires some more manual work~

**Update:** Multi-series pie charts are now functional too

**Comment by joepavitt** [comment]
~I've also matched appearances across the board, so it shouldn't be too apparent that updates have occurred. The only primary difference, which again, can be done with some manual work, is the default empty state - looks much nicer with ChartJS (which uses a default `0` > `1` y-axis) than in eCharts (has no grid or values on the respective axes)~

**Update:** I've now implemented a manual method to show a default 0 > 1 y-axis until we receive real data.

**Comment by joepavitt** [comment]
~Just realised that I've broken the trimming (it was intentional whilst migrating, just forgot to re-introduce/write it), that's now the next task here~

**Update:** Fixed

**Comment by joepavitt** [comment]
Noticed https://github.com/FlowFuse/node-red-dashboard/issues/1811 popup, and given I'd just managed to get the migration completed, thought I'd give it a quick go, and it was successful, as such I've included it in the incoming PR

https://github.com/user-attachments/assets/a2560e71-ce3e-479e-ae93-250be77f6018

**Comment by joepavitt** [comment]
~One final sweep of testing, found an issue when using keymapping. Not as simple in eCharts as in ChartJS. Investigating.~

**Update:** Was a red herring and the problem was actually in the flow, not the code.

**Comment by joepavitt** [comment]
Last last step - update the E2E tests.

**Comment by colinl** [comment]
I have noticed a rather odd symptom.  It is visible in the top chart in the video you posted a couple of posts ago.  When new data points are added the vertical grid lines sometimes become much closer together at the right hand side.

**Comment by colinl** [comment]
It can happen at the left hand side too.  It seems that it fits a fixed number of divisions between each timestamp marker, and for the leftmost and rightmost markers it fits that number of divisions between the marker and the end of the chart.

**Comment by joepavitt** [comment]
> When new data points are added the vertical grid lines sometimes become much closer together at the right hand side.

Yeah, I haven't quite worked out the _why_ for that, it's fully functional, and the labels are correct, but for some reason eCharts decides it needs more minor gridlines in place.

I'll get an issue raised for it, but certainly low priority and not a blocker for us shipping the migration. Thanks for flagging @colinl

**Comment by colinl** [comment]
I agree, not a blocker.

**Comment by colinl** [comment]
I have done some further testing with various configurations and haven't found anything else other than the point I noted in #1811 (categorical x axis for line chart).

**Comment by joepavitt** [comment]
Have reduced this to a size 5 to reflect the actual size in the end.
